### ISSUE TITLE
installing keychain on opensuse results in black screen after boot

### ISSUE THREAD

**Issue by guillermo-st** [funtoo/keychain#147]
I'm trying to set up keychain on opensuse to automatically add my github ssh keys on boot, and prevent the password input popups on every commit. After installing keychain and adding my keys, it works just fine until the next reboot. After turning the machine on, opensuse gets hung up on a black screen. The only workaround is to roll the OS back to a previous snapshot where keychain was not installed.

I have even tried reinstalling opensuse from scratch (including a full disk format), but this only happens after installing keychain. This is the line I added to my .bashrc:

`eval \`keychain --eval $HOME/.ssh/* \` `

I have no idea how to fix this, I guess I will have to settle for manually starting the ssh agent myself. Help much appreciated!

**Comment by oversize** [comment]
I dont think you need that $HOME/.ssh/* thing. If anything you should specify the actual key you want to load (and not everythin inside .ssh/). E.g. `id_rsa id_ed25519`

Also, you might want to use $() instead of backticks. 
Like so `eval $(keychain --eval )` works fine with multiple keys for me ... on debian though.

**Comment by dominicf2001** [comment]
This is occurring on Fedora KDE spin as well. It happens right after I log in.

In my `~/.bashrc`

`eval $(keychain --eval --quiet id_ed25519)`

**Comment by danielrobbins** [comment]
I think this is somewhat odd. I would specify the keys just by their name, not by their path, and get rid of the wildcard.

Let's get more information. What is the version of keychain, and did you install it from the distro or manually? Distributions could add patches to keychain.

Please try keychain-2.9.0_beta3 or later from GitHub and see if this problem persists, as this version has been rewritten and it would be good to know if it is still a problem.

Beyond this, more troubleshooting is needed to identify the root cause of the issue so we know what exactly is going on. When the system hangs, try to ssh in remotely, get output of "ps aux --forest" which should show if any processes have been launched by keychain that are waiting for input, so should give us more information.

**Comment by JamesB192** [comment]
Huh, it looks like I am sort of right for aarch64 Fedora 42 in UTM.

A clause like the following remedies it after determining that neither DISPLAY nor WAYLAND_DISPLAY was being set at that point.

```sh
if test -n "$DISPLAY" -o -n "$WAYLAND_DISPLAY"; then
    eval $(keychain --eval --systemd id_ed25519)
fi
```

However, this requires another Bash invocation, like starting Konsole.

**Comment by redvelociraptor** [comment]
I got a black screen as well on two distros using KDE & Wayland when I added keychain to .zshenv. SDDM pulls in .zshenv at GUI login, but not .zshrc, so I moved keychain to .zshrc and check if it's an interactive shell before invoking it.

**Comment by danielrobbins** [comment]
Thanks for the report and the follow-up notes here. I’m going to close this as a legacy 2.x / shell-startup placement issue rather than an active Keychain 3.x bug.

The later comments point to Keychain being invoked from files that can be sourced by the display manager or non-interactive GUI startup paths, such as `.zshenv`, before a usable interactive/display environment exists. The recommended setup is to run Keychain from an interactive/login shell startup file such as `.zshrc`, `.bashrc`, or `.bash_profile`, with an interactive-shell guard if needed.

Keychain 3.x also now has `keychain inspect` to make this kind of environment troubleshooting much easier. If this still reproduces with Keychain 3.x from an appropriate interactive shell startup path, please open a fresh issue with the exact shell startup snippet and `keychain inspect` output.
